icu4c/source/common/ulocbuilder.cpp

```diff
@@ -0,0 +1,145 @@
+ULocaleBuilder* ulocbld_open() {
+    return EXTERNAL(new icu::LocaleBuilder());
+}
+
+void ulocbld_close(ULocaleBuilder* builder) {
+    if (builder == nullptr) return;
+    delete INTERNAL(builder);
+}
+
+void ulocbld_setLocale(ULocaleBuilder* builder, const char* locale, int32_t length) {
+    if (builder == nullptr) return;
+    icu::Locale l;
+    if (length < 0 || locale[length] == '\0') {
+        l = icu::Locale(locale);
+    } else {
+        if (length >= ULOC_FULLNAME_CAPACITY) {
+            l.setToBogus();
+        } else {
+            // locale is not null termined but Locale API require one.
+            // Create a null termined version in buf.
+            char buf[ULOC_FULLNAME_CAPACITY];
+            uprv_memcpy(buf, locale, length);
+            buf[length] = '\0';
+            l = icu::Locale(buf);
+        }
+    }
+    INTERNAL(builder)->setLocale(l);
+}
+
+#define STRING_PIECE(s, l) ((l)<0 ? StringPiece(s) : StringPiece((s), (l)))
+
+#define IMPL_ULOCBLD_SETTER(N) \
+void ulocbld_##N(ULocaleBuilder* bld, const char* s, int32_t l) { \
+    if (bld == nullptr) return; \
+    INTERNAL(bld)->N(STRING_PIECE(s,l)); \
+}
+
+IMPL_ULOCBLD_SETTER(setLanguageTag)
+IMPL_ULOCBLD_SETTER(setLanguage)
+IMPL_ULOCBLD_SETTER(setScript)
+IMPL_ULOCBLD_SETTER(setRegion)
+IMPL_ULOCBLD_SETTER(setVariant)
+IMPL_ULOCBLD_SETTER(addUnicodeLocaleAttribute)
+IMPL_ULOCBLD_SETTER(removeUnicodeLocaleAttribute)
+
+void ulocbld_setExtension(ULocaleBuilder* builder, char key, const char* value, int32_t length) {
+    if (builder == nullptr) return;
+    INTERNAL(builder)->setExtension(key, STRING_PIECE(value, length));
+}
+
+void ulocbld_setUnicodeLocaleKeyword(
+    ULocaleBuilder* builder, const char* key, int32_t keyLength,
+    const char* type, int32_t typeLength) {
+    if (builder == nullptr) return;
+    INTERNAL(builder)->setUnicodeLocaleKeyword(
+        STRING_PIECE(key, keyLength), STRING_PIECE(type, typeLength));
+}
+
+void ulocbld_clear(ULocaleBuilder* builder) {
+    if (builder == nullptr) return;
+    INTERNAL(builder)->clear();
+}
+
+void ulocbld_clearExtensions(ULocaleBuilder* builder) {
+    if (builder == nullptr) return;
+    INTERNAL(builder)->clearExtensions();
+}
+
+int32_t ulocbld_buildLocaleID(ULocaleBuilder* builder,
+                              char* buffer, int32_t bufferCapacity, UErrorCode* err) {
+    if (builder == nullptr) {
+        *err = U_ILLEGAL_ARGUMENT_ERROR;
+        return 0;
+    }
+    icu::Locale l = INTERNAL(builder)->build(*err);
+    if (U_FAILURE(*err)) return 0;
+    int32_t length = (int32_t)(uprv_strlen(l.getName()));
+    if (length <= bufferCapacity) {
+        uprv_strncpy(buffer, l.getName(), length);
+        if (length < bufferCapacity) {
+            buffer[length] = '\0';
+        } else {
+            *err = U_STRING_NOT_TERMINATED_WARNING;
+        }
+        return length;
+    }
+    *err = U_BUFFER_OVERFLOW_ERROR;
+    uprv_memcpy(buffer, l.getName(), bufferCapacity);
+    return length;
```

Comment:
@FrankYFTang @richgillam sorry for not paying attention for way too long...
Any C API that writes a string to a destination buffer should return with a call to [u_terminateChars()](https://github.com/unicode-org/icu/blob/main/icu4c/source/common/ustr_imp.h#L81), u_terminateUChars(), ... as appropriate, and those functions handle the errors and warnings and NUL-termination. See other C API implementations. Please send a follow-up PR.

Reply:
@markusicu see https://github.com/unicode-org/icu/pull/2565 for the follow-up PR